Get timezone and menu button

KaterinaSolovyeva opened the pull request:
Ticket - [Установка дефолтных кнопок для телеграма](https://www.notion.so/7d6d35f407814742a0cebea1e480bd4b?p=9f3edd440b0a4ffca5d77f5e695a2870)

добавлена кнопка Меню, из которой можно перейти в список доступных запросов

Ticket - [Установка часового пояса + диалоги бота к нему](https://www.notion.so/7d6d35f407814742a0cebea1e480bd4b?p=599621be7b1f47e19470172398a5db7e)

При запросе на присвоение нового часового пояса появляются две кнопки объекта ReplyKeyboard:
'Отправить свою геолокацию для автоматической настройки часового пояса' - получает гелолокацию, по ней определяет таймзону. Передает по апи
'Напишу свою таймзону сам' - распознает введенный текст. Проверяет, существует ли такая таймзона. Если да, то передает её через апи. Если нет, то проверяет, не является ли введенный текст тем, на который отвечать не нужно. Просит пользователя ввести другую таймзону или выслать геолокацию.
После получения таймзоны кнопки ReplyKeyboard внизу исчезают

IvanOsyanin committed:
logger added

IvanOsyanin committed:
1st review changes

IvanOsyanin committed:
Merge branch 'develop' into feature/logging

IvanOsyanin committed:
logging changes

Danielseddian committed:
Add trello endpoint and base function for webhook

Danielseddian committed:
Add trello endpoint and base function for webhook

Danielseddian committed:
Add trello endpoint and base function for webhook

Danielseddian committed:
Add trello endpoint and base function for webhook

Danielseddian committed:
Add trello endpoint and base function for webhook

Danielseddian committed:
Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

zhss1983 committed:
Changed the names of constants and their values where necessary.

agamova committed:
Dockerfile support added

zhss1983 committed:
Combined the bot_menu and start_conversation files. Replaced the "start" command with a constant that was previously created, but was not used in the project.

zhss1983 committed:
Highlighted unused constants.

zhss1983 committed:
I moved the constants from the project to a separate folder. I didn't make any changes, the status can be viewed by the previous commit.

AndreyMurysev committed:
add new models

AndreyMurysev committed:
add new models

AndreyMurysev committed:
add new models

zhss1983 committed:
Deleted unused constants

zhss1983 committed:
Moved callbacks to callback_data file

zhss1983 committed:
Add cansel command to constants

zhss1983 committed:
Restored constants folder and moved URL_SERVICE_RULES to enviroments

zhss1983 committed:
Removed cansel_comand and it's methods.

zhss1983 committed:
Removed cansel_comand and it's methods. Fix PEP8 error

zhss1983 committed:
Rename callback constants to command ones.

zhss1983 committed:
Fix REPEAT_COMMAND constant

zhss1983 committed:
Fix REPEAT_COMMAND constant

zhss1983 committed:
Изменил имена констант на CALLBACK_<имя метода который он вызывает>_COMMAND для того чтобы сразу становилось ясно для чего нужна та или иная константа.
Перенёс запуск меню в start_conversation по команде /menu.

zhss1983 committed:
I changed the names of the constants to CALLBACK_<the name of the method that it calls>_COMMAND so that it immediately becomes clear what this or that constant is needed for.
Moved the menu launch to start_conversation by the /menu command.
I saved callback_menu_handler in case it is imported in the future, the code is not mine, I do not know how it will be reused.

agamova committed:
start method changed

IvanOsyanin committed:
Merge branch 'develop' into feature/logging

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

agamova committed:
gunicorn removed

agamova committed:
added comment to Dockerfile

AndreyMurysev committed:
del api

AndreyMurysev committed:
ppull develop

kapkaevandrey committed:
Update function, add new argument parse_mode

AndreyMurysev committed:
rename files service

AndreyMurysev committed:
DESIGN pylint add max-attributes=9

kapkaevandrey committed:
Fix spelling mistakes.

kapkaevandrey committed:
Merge remote-tracking branch 'origin/develop' into develop

kapkaevandrey committed:
Add function to mailing send

kapkaevandrey committed:
Add template messages.

kapkaevandrey committed:
Add logging when request statistic

kapkaevandrey committed:
Style Fix

kapkaevandrey committed:
add methods to jobs

kapkaevandrey committed:
add to job TODO for logic when mailing Failed

kapkaevandrey committed:
Add docstring

Danielseddian committed:
Updated healthcheck_api and trello_webhook_api.

Danielseddian committed:
Updated healthcheck_api and trello_webhook_api.

Danielseddian committed:
Updated healthcheck_api and trello_webhook_api.

Danielseddian committed:
Updated healthcheck_api and trello_webhook_api.

kapkaevandrey committed:
Fix style mistakes

KaterinaSolovyeva committed:
Removed the excess from PR

KaterinaSolovyeva committed:
Removed the excess from PR

KaterinaSolovyeva committed:
Removed the excess from PR

KaterinaSolovyeva committed:
Removed the excess from PR

MatMerd commented on src/get_timezone.py:
```diff
@@ -0,0 +1,81 @@
+    else:
+        text = text.replace(" ", "")
+        try:
+            utc = int(list(filter(None, text.split("UTC")))[0])
```
Такие однострочники это конечно прикольно, но в действительности они чаще всего сложно поддерживаемые.
Почему нельзя сделать просто int(text.lstrip("UTC"))?
Да и вообще, "UTC" - это хардкод и лучше это выносить в константы, чтоб мы в любой момент могли поменять.
И ещё момент. На вики можно увидеть, что таймзоны бывают не только целыми, но и +3:30 например. С этим тоже надо как-то работать.

KaterinaSolovyeva replied:
может, тогда передавать не часы, а минуты по апи?

MatMerd replied:
Нет. Можно передавать часы дробью. 3.5 - это 3:30. 3.75 - 3:45.

KaterinaSolovyeva replied:
ок

veselij replied:
я извиняюсь, я просил изначально тайм-зону числом. давайте оставим пока str "UTC+3.75". и так и будем сохранять в БД, там тоже сейчас str. а потом в функции расчета времени отправки по ТЗ будем уже высчитывать

KaterinaSolovyeva replied:
хорошо

KaterinaSolovyeva replied:
такие выводы подойдут или обязательно с плюсом? 
UTC09:30
UTC-3:00

AntonZelinsky replied:
Ой, а что за таймзона `UTC+3.75`? Первый раз слышу, да и гугл такую не нашёл! И на википедии такой статьи не нашёл..(

Предлагаю написать регулярку проверяющую на два формата `UTC+3` и просто `+5`. `UTC` не хочется заставлять добавлять потому что нужно будет переключаться каждый раз на другую клавиатуру и велики шансы ошибок, которые нужно уметь обрабатывать.
Что делать, если кто-то всё  же додумается написать +3.5? Я считаю, что нужно отбросить то, что после точки или запятой и всё.

AntonZelinsky replied:
+ обязателен

KaterinaSolovyeva replied:
Australia/Darwin +9:30

KaterinaSolovyeva replied:
https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

MatMerd replied:
Я уже думал про регулярку. Но если таймзона составная, регулярка становится сильно не удобной. И тоже. Писать её локально в коде - не круто. Хапдкод получается

KaterinaSolovyeva replied:
Вопросы: 
1. что в итоге передавать?
Так: UTC+02:00, UTC-02:30
Или так: UTC+2, UTC-2.5?
2. что просим ввести пользователя?
Так: UTC+02:00, UTC-02:30, UTC+2, +2, -2:30 ?
Или иначе?

Уточните задачу, пожалуйста

AntonZelinsky replied:
Ой, мне кажется,  тут регулярка будет очень простой. 
Матвей, подскажи кейсы которые тебя смущают?

veselij committed:
Merge pull request #25 from Studio-Yandex-Practicum/feature/add_trello_end_point

Feature/add trello end point

kapkaevandrey committed:
Replace send_statistic function to the core -> send_message.py

kapkaevandrey committed:
Refactoring send mailing.

kapkaevandrey committed:
Delete logic for failed mailing.

kapkaevandrey committed:
the logic of statistics collection and message preparation has been moved

kapkaevandrey committed:
Rename function.

kapkaevandrey committed:
Fix style mistakes

zhss1983 committed:
Feature/pre commit (#33)

* Исправил README.md - подправл строку 'pip install -r ...'

* Added errors E0401 and R0902 to ignore list

* Undo readme file changes.

* Added error R0902 to ignore list.

* Исправил README.md - подправл строку 'pip install -r ...'

* Added errors E0401 and R0902 to ignore list

* Returned readme to correct state

* Changed the names of constants and their values where necessary.

* Combined the bot_menu and start_conversation files. Replaced the "start" command with a constant that was previously created, but was not used in the project.

* Highlighted unused constants.

* I moved the constants from the project to a separate folder. I didn't make any changes, the status can be viewed by the previous commit.

* Deleted unused constants

* Moved callbacks to callback_data file

* Add cansel command to constants

* Restored constants folder and moved URL_SERVICE_RULES to enviroments

* Removed cansel_comand and it's methods.

* Removed cansel_comand and it's methods. Fix PEP8 error

* Rename callback constants to command ones.

* Fix REPEAT_COMMAND constant

* Fix REPEAT_COMMAND constant

* Изменил имена констант на CALLBACK_<имя метода который он вызывает>_COMMAND для того чтобы сразу становилось ясно для чего нужна та или иная константа.
Перенёс запуск меню в start_conversation по команде /menu.

* I changed the names of the constants to CALLBACK_<the name of the method that it calls>_COMMAND so that it immediately becomes clear what this or that constant is needed for.
Moved the menu launch to start_conversation by the /menu command.
I saved callback_menu_handler in case it is imported in the future, the code is not mine, I do not know how it will be reused.

* Dockerfile support added

* start method changed

* gunicorn removed

* added comment to Dockerfile

* Rebased with develop branch.

* Solved problems W0611, W1203, E0401 and E1101.

* Solved problems R0902.

* Update bot.py

Fixed PEP8 problem.

* Removed src. and __init__.py

* # pylint: disable=R0902

* Update config.py

Removed CHAT_ID

* callback -> callback_data

Co-authored-by: Evgenia Agamova <evgenia@agamova.ru>

agamova committed:
webhook url changed and added to env.example

veselij committed:
Merge pull request #36 from Studio-Yandex-Practicum/fix/init_webhook

webhook url changed and added to env.example

agamova committed:
env..example changed

kapkaevandrey committed:
All logic relocate to jobs.py

kapkaevandrey committed:
add fake id

kapkaevandrey committed:
Merge branch 'develop' into feature/sending_statistics

IvanOsyanin committed:
review changes

IvanOsyanin committed:
Merge branch 'develop' into feature/logging

kapkaevandrey committed:
Added mapping username -> user_id

kapkaevandrey committed:
Added logging for situation when username not in bot_data.

KaterinaSolovyeva committed:
Merge branch 'feature/fixed_menu_button_timezone' of https://github.com/Studio-Yandex-Practicum/ask_nenaprasno_bot into feature/fixed_menu_button_timezone

KaterinaSolovyeva committed:
Deleted get timezone from text code

KaterinaSolovyeva committed:
Deleted get timezone from text code

KaterinaSolovyeva committed:
Deleted get timezone from text code

AndreyMurysev committed:
pylint

AndreyMurysev committed:
upgrade site_api_service

veselij committed:
Merge pull request #37 from Studio-Yandex-Practicum/fix/env_example

env..example changed

kapkaevandrey committed:
Relocate create dict with mapping useranme to user id in bot create.

K3llar committed:
refactor dependencies in pyproject.toml

kapkaevandrey committed:
Added a check that the username_to_id key is present in the bot_data dictionary

AndreyMurysev committed:
upgrade site_api_service

kapkaevandrey committed:
Little refactoring, relocate all logic in send_message.py

kapkaevandrey committed:
Style fix

AndreyMurysev committed:
Merge branch 'develop' into feature/refactoring_the_service_API

AndreyMurysev committed:
delete docstring in SiteAPIService

AndreyMurysev committed:
pull develop

AntonZelinsky committed:
Merge pull request #32 from Studio-Yandex-Practicum/feature/refactoring_the_service_API

Feature/refactoring the service api

kapkaevandrey committed:
Merge remote-tracking branch 'origin/develop' into develop

kapkaevandrey committed:
Merge branch 'develop' into feature/sending_statistics

kapkaevandrey committed:
rename telegram_name(str) to telegram_id(int) fix little mistakes

kapkaevandrey committed:
rename telegram_name(str) to telegram_id(int) fix little mistakes.

kapkaevandrey committed:
Change authorize from telegram_name to telegram_id. Delete bot_data. Delete logging from send_statistic. Update jobs. Work test status - success.

kapkaevandrey committed:
Fix style mistakes.

kapkaevandrey committed:
Full delete bot_data)

kapkaevandrey committed:
Fix Dockerfile paradox.

kapkaevandrey committed:
Fix README.md paradox.

kapkaevandrey committed:
Fix run_webhook_api.py paradox.

kapkaevandrey committed:
Renamed attribute telegram_name in method set_user_timezone to telegram_id

kapkaevandrey committed:
Merge pull request #34 from Studio-Yandex-Practicum/feature/sending_statistics

Feature/sending statistics

veselij committed:
Merge pull request #39 from Studio-Yandex-Practicum/feature/refactor_dependencies

IvanOsyanin committed:
Merge pull request #16 from Studio-Yandex-Practicum/feature/logging

logger added

kapkaevandrey committed:
Rename telegram_name to telegram_id also fix annotation.

kapkaevandrey committed:
Style Fix

kapkaevandrey committed:
Another stupid mistake.

KaterinaSolovyeva committed:
Merging

kapkaevandrey committed:
Merge pull request #43 from Studio-Yandex-Practicum/fix_api_name_attribute

Rename telegram_name to telegram_id also fix annotation.

KaterinaSolovyeva committed:
Merge branch 'feature/fixed_menu_button_timezone' of https://github.com/Studio-Yandex-Practicum/ask_nenaprasno_bot into feature/fixed_menu_button_timezone

veselij commented on src/get_timezone.py:
```diff
@@ -7,6 +7,8 @@
 
 from service.api_client_fake import FakeAPIService
```
там сейчас поменялось после рефакторинг

veselij commented on src/get_timezone.py:
```diff
@@ -26,14 +28,14 @@ async def get_timezone(update: Update, context: CallbackContext) -> None:
+        text=f"Установлен часовой пояс для {text_utc}",
         reply_markup=ReplyKeyboardRemove(),
     )
     api = FakeAPIService()
```
надо теперь использовать просто api_service = APIService()

veselij commented on src/get_timezone.py:
```diff
@@ -43,37 +45,27 @@ async def get_timezone_from_location(update: Update, context: CallbackContext) -
         utc_time = datetime.datetime.utcnow()
-        utc = int(time_zone.utcoffset(utc_time).total_seconds() // 3600)
-        await set_timezone(update.effective_chat.id, utc, context)
+        utc = float(time_zone.utcoffset(utc_time).total_seconds() / 3600)
+        hours, minutes = divmod(utc * 60, 60)
```
тут какая-то магия, не могу понять, ты же уже перевела в часы, а потом опять в минуты.

KaterinaSolovyeva replied:
мне нужны часы и минуты, чтобы записать в виде hours:minutes.
Мне показалось, что это самый простой и наглядный способ
строчка 53 высчитывает сколько часов(не округляя) например: 3.15
54 - разбивает это значение на часы и минуты. Например 3 часа 9 минут

KaterinaSolovyeva replied:
сделать так?:
```
utc = float(time_zone.utcoffset(utc_time).total_seconds() / 60)
        hours, minutes = divmod(utc, 60)
```

MatMerd replied:
Ну по факту , это действительно самый лёгкий способ

MatMerd replied:
@KaterinaSolovyeva да, сделай так, как указала в последнем сообщении

veselij commented on src/get_timezone.py:
```diff
@@ -43,37 +45,27 @@ async def get_timezone_from_location(update: Update, context: CallbackContext) -
-                        " определения."
-                    ),
-                )
+        pass
```
давай сюда место pass напишем в чат, что "вы установили таймзону X"

KaterinaSolovyeva committed:
Changed apiservice

KaterinaSolovyeva committed:
Changed pass to sending text
